src/Control/Director.php

```diff
@@ -821,7 +821,11 @@ public static function is_site_url($url)
 
         // Allow extensions to weigh in
         $isSiteUrl = false;
-        static::singleton()->extend('updateIsSiteUrl', $isSiteUrl, $url);
+        // not using static::singleton() here because it can, for unknown reasons, break
+        // functional tests such as those in HTTPCacheControlIntegrationTest
+        // extension hooks will still be called as expected
+        $director = new static();
+        $director->extend('updateIsSiteUrl', $isSiteUrl, $url);
```

Comment:
This screamed "underlying bug" to me so I've looked into it a bit and found the cause.

Instantiating a singleton includes applying the middleware, which I think is ultimately the cause of the failure - but regardless of what it is about the singleton that causes it, the main problem is a singleton of `Director` being instantiated before the test is run.

`Controller::normaliseTrailingSlash()` is called during `SapphireTest::setUp()`
1. `HTTPCacheControlIntegrationTest::setUp()`
1. `FunctionalTest::setUp()`
1. `SapphireTest::setUp()`
1. `SapphireTest::setUpRoutes()`
1. `SapphireTest::getExtraRoutes()`
1. `RequestHandler::Link()`
1. `Controller::join_links()`
1. `Controller::normaliseTrailingSlash()`

We should probably add this to the end of `SapphireTest::setUp()`:

```php
Injector::inst()->unregisterNamedObject(Director::class);
```

This will protect us from similar regressions in the future.

Alternatively, please at least update the comment to reflect that the problem is caused by instantiating a singleton of `Director` prior to tests being run, and that for this specific use case we don't need to use dependency injection because the chance of the `extend()` method being customised in projects is low.

Reply:
Have gone for the better comment comment